talk/owt/sdk/p2p/p2ppeerconnectionchannel.cc

```diff
@@ -163,6 +163,121 @@ void P2PPeerConnectionChannel::ClearPendingStreams() {
   }
 }
 
+void P2PPeerConnectionChannel::PublishBatch(std::vector<std::shared_ptr<LocalStream>> streams,
+    std::function<void(std::shared_ptr<LocalStream>)> on_success,
+    std::function<void(std::unique_ptr<Exception>)> on_failure)
+{
+  // Add reference to peer connection until end of function.
+  rtc::scoped_refptr<webrtc::PeerConnectionInterface> temp_pc_ = GetPeerConnectionRef();
+
+  if (!temp_pc_) {
+    RTC_LOG(LS_INFO) << "Peer connection closed, returning.";
+    // Skip on_success callback because we haven't published anything.
+    ClearPendingStreams();
+    if (on_failure) {
+      std::unique_ptr<Exception> e(
+          new Exception(ExceptionType::kP2PClientRemoteNotExisted,
+                        "Peer connection closed."));
+      on_failure(std::move(e));
+    }
+    return;
+  }
+
+  auto stream_labels = std::vector<std::string>();
+  std::for_each(streams.begin(), streams.end(), [&stream_labels](std::shared_ptr<LocalStream> stream) {
+    if (stream) {
+      // Calling [stream|track]->id() runs on signaling_thread, so call outside of locks
+      stream_labels.push_back(stream->MediaStream()->id());
+    } else {
+      RTC_LOG(LS_INFO) << "Local stream cannot be nullptr. Skipping.";
+    }
+  });
+
+  {
+    std::lock_guard<std::mutex> lock(published_streams_mutex_);
+    for (const auto &stream_label : stream_labels) {
+      if (published_streams_.find(stream_label) != published_streams_.end() ||
+          publishing_streams_.find(stream_label) != publishing_streams_.end()) {
+        if (on_failure) {
+          std::unique_ptr<Exception> e(
+              new Exception(ExceptionType::kP2PClientInvalidArgument,
+                            "The stream is already published."));
+          on_failure(std::move(e));
+        }
+      }
+    }
+  }
+
+  // Prune null streams
+  streams.erase(std::remove_if(streams.begin(), streams.end(), [](std::shared_ptr<LocalStream> stream) {
+    return !stream;
+  }), streams.end());
+
+
+  for (auto stream : streams) {
+    RTC_LOG(LS_INFO) << "Publishing a local stream.";
+
+    // Send chat-closed to workaround known browser bugs, together with
+    // user agent information once and once only.
+    {
+      std::lock_guard<std::mutex> lock(stop_send_mutex_);
+      if (stop_send_needed_) {
+        SendStop(nullptr, nullptr);
+        stop_send_needed_ = false;
+      }
+      if (!ua_sent_) {
+        SendUaInfo();
+        ua_sent_ = true;
+      }
+    }
```

Comment:
this whole block can be brought outside the for loop. it will run at most once

Reply:
I just wasn't sure of doing that as semantically the order in which we do things will be different now. i.e. currently we send this message once, and then we call the success callback for the stream. I suppose we can move it around, as we're not sending any other messages in the callback.